R/utils-plumber.R

```diff
@@ -222,3 +222,124 @@ extract_endpoint <- function(path) {
   # stringr::str_extract(path, pattern = "([^/]+$)")
   sub(".*[/]", "", path)
 }
+
+
+#' Return the version of the data
+#'
+#' @param version Data version. Defaults to most recent version. See api/v1/versions
+#' @param release_version date when the data was published in YYYYMMDD format
+#' @param ppp_version ppp year to be used
+#' @param identity One of "PROD" (production), "INT" (internal) and "TEST"
+#' @param versions_available character vector of all the versions available
+#'
+#' @return character
+#'
+#' @export
+#'
+return_correct_version <- function(version = NULL,
+                                   release_version = NULL,
+                                   ppp_version = NULL,
+                                   identity = 'PROD',
+                                   versions_available) {
+  # If version is passed return it directly.
+  if (!is.null(version)) return(version)
+
+  if(!is.null(release_version) && !is.null(ppp_version) && !is.null(identity)) {
+    selected_version <- rpi_version(release_version, ppp_version, identity, versions_available)
+  } else if(!is.null(release_version) && !is.null(ppp_version)) {
+    #This probably would never be executed since identity would never be NULL unless explicitly specified.
+    selected_version <- rp_version(release_version, ppp_version, versions_available)
+  } else if(!is.null(release_version) && !is.null(identity)) {
+    selected_version <- ri_version(release_version, identity, versions_available)
+  } else if(!is.null(ppp_version) && !is.null(identity)) {
+    selected_version <- pi_version(ppp_version, identity, versions_available)
+  }
+  #If no matching version is found
+  if (length(selected_version) == 0)
+    #Since the function returns character values
+    return("404")
+  #If only 1 value matches
+  else if (length(selected_version) == 1)
+    return(selected_version)
+  #If more than 1 value matches
+  #If release_version is not null get the max version from ppp date
+  if (!is.null(release_version)) return(select_max_version_from_ppp(selected_version))
+  else return(select_max_version_from_release_version(selected_version))
```

Comment:
If I remember correctly, the versioning nomenclature works in a way such that the following code always returns the correct version. In other words, there is no need for this "if-else" conditions. Could you please confirm? 

Thanks. 

```r
selected_version[which.max(selected_version)]
```

Reply:
Maybe you are right. However, `selected_version` is a string (Eg - `20220609_2011_02_02_PROD`) and `which.max` is not recommended to use on string values.

Reply:
Fair enough. I actually don't use `which.max` but `max`, which works well with characters. So, this is more than enough, `max(selected_version)`. Since the characters we use in the versioning are relatively standard, and do not depend on any particular locale,  I think we can use `max` safely. Please, try it out and let me know how it goes. 

Thanks!

Reply:
I thought about this and test on few examples and I think you are right. Using only `max` should be sufficient here. I made the changes accordingly. Thank you for the feedback.

Reply:
Hey - What about this:
```
tmp <- c("20221212_2017_03_01_PROD", "20221212_2017_03_01_TEST")
> max(tmp)
[1] "20221212_2017_03_01_TEST"
```

It looks like the suffix can be problematic for returning the correct version here, no? If a *_TEST folder makes it to the PROD server by mistake, this could be very problematic. What do you think?

Even if the `select_max_version_from_release_version` function is just a thin wrapper around `max(selected_version)`, I would rather have it as a separate function so we can unit test it for such edge cases. I'm reviewing this PR now, and will add the function back along with it's unit tests.

Reply:
but I can't think of scenario where we will have version from more than one `identity` in `selected_version`. We always have an `identity` argument with default value as `PROD` so even if `_TEST` folder makes it to PROD it will never be selected unless you explicitly specify `identity=TEST`.